pkg/skaffold/build/parallel.go

```diff
@@ -21,96 +21,103 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"sync"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/tag"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
-	"github.com/pkg/errors"
 )
 
 const bufferedLinesPerArtifact = 10000
 
 type artifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
 
 // InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact artifactBuilder) ([]Artifact, error) {
+func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact artifactBuilder) ([]chan Result, error) {
 	if len(artifacts) == 1 {
 		return InSequence(ctx, out, tags, artifacts, buildArtifact)
 	}
 
+	var resultChans []chan Result
+
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	n := len(artifacts)
-	finalTags := make([]string, n)
-	errs := make([]error, n)
-	outputs := make([]chan []byte, n)
+	// for collecting all output and printing in order later on
+	outputs := make([]chan []byte, len(artifacts))
+
+	for i, a := range artifacts {
+		resChan := make(chan Result, 1)
+		resultChans = append(resultChans, resChan)
 
-	// Run builds in //
-	for index := range artifacts {
-		i := index
+		// give each build a byte buffer to write its output to
 		lines := make(chan []byte, bufferedLinesPerArtifact)
 		outputs[i] = lines
 
-		r, w := io.Pipe()
-
-		// Log to the pipe, output will be collected and printed later
-		go func() {
-			// Make sure logs are printed in colors
-			var cw io.WriteCloser
-			if color.IsTerminal(out) {
-				cw = color.ColoredWriteCloser{WriteCloser: w}
-			} else {
-				cw = w
+		go func(artifact *latest.Artifact, c chan Result, lines chan []byte) {
+			res := &Result{
+				Target: *artifact,
 			}
-
-			color.Default.Fprintf(cw, "Building [%s]...\n", artifacts[i].ImageName)
-
-			event.BuildInProgress(artifacts[i].ImageName)
-
-			tag, present := tags[artifacts[i].ImageName]
-			if !present {
-				errs[i] = fmt.Errorf("unable to find tag for image %s", artifacts[i].ImageName)
-				event.BuildFailed(artifacts[i].ImageName, errs[i])
-			} else {
-				finalTags[i], errs[i] = buildArtifact(ctx, cw, artifacts[i], tag)
-				if errs[i] != nil {
-					event.BuildFailed(artifacts[i].ImageName, errs[i])
+			wg := &sync.WaitGroup{}
+			wg.Add(2) // one for the build, one for the output processing
```

Comment:
I'd move Add(1) in front of each of the `go func()`'s

Reply:
k